File: services/python-service/services/utils/webhook_client.js
import axios from 'axios';

const API_GATEWAY_URL = process.env.API_GATEWAY_URL || 'http://localhost:3001';

const webhookClient = axios.create({
    baseURL: API_GATEWAY_URL,
    timeout: 10000, // 10 Sekunden Timeout für Webhooks
    headers: {
        'Content-Type': 'application/json'
    }
});

/**
 * Sendet einen Webhook an den API Gateway, wenn ein Job abgeschlossen ist
 * @param {string} jobId - ID des Jobs
 * @param {string} jobType - Typ des Jobs (training, retraining, prediction)
 * @param {string} projectId - ID des Projekts
 * @param {Object} result - Ergebnis des Jobs
 * @param {string} status - Status (completed, failed)
 */
export async function sendJobCompletionWebhook(jobId, jobType, projectId, result, status) {
    try {
        const webhookData = {
            jobId,
            jobType,
            projectId,
            result,
            status
        };

        console.log(`Sende Webhook für Job ${jobId} (${jobType}) an API Gateway...`);

        const response = await webhookClient.post('/api/webhooks/job-completed', webhookData);

        console.log(`Webhook für Job ${jobId} erfolgreich gesendet`);
        return response.data;
    } catch (error) {
        // Webhook-Fehler sollten nicht den Job-Status beeinflussen
        console.error(`Fehler beim Senden des Webhooks für Job ${jobId}:`, error.message);
        console.error(`Webhook-URL: ${API_GATEWAY_URL}/api/webhooks/job-completed`);

        // Versuche es erneut nach kurzer Verzögerung (optional)
        if (error.code === 'ECONNREFUSED' || error.code === 'ETIMEDOUT') {
            console.warn(`API Gateway nicht erreichbar, versuche Webhook erneut in 5 Sekunden...`);
            setTimeout(async () => {
                try {
                    await webhookClient.post('/api/webhooks/job-completed', {
                        jobId,
                        jobType,
                        projectId,
                        result,
                        status
                    });
                    console.log(`Webhook für Job ${jobId} erfolgreich nach Wiederholung gesendet`);
                } catch (retryError) {
                    console.error(`Webhook-Wiederholung für Job ${jobId} fehlgeschlagen:`, retryError.message);
                }
            }, 5000);
        }

        // Wir werfen den Fehler nicht, damit der Job trotzdem als abgeschlossen gilt
        return null;
    }
}

export default {
    sendJobCompletionWebhook
};

File: services/unified-service/infrastructure/clients/webhook_client.py
"""
Webhook-Client für API Gateway
"""

import os
import requests
import time
from typing import Dict, Any, Optional

API_GATEWAY_URL = os.getenv('API_GATEWAY_URL', 'http://localhost:3001')

def send_job_completion_webhook(
    job_id: str,
    job_type: str,
    project_id: str,
    result: Optional[Dict[str, Any]],
    status: str
) -> Optional[Dict[str, Any]]:
    """Sendet Webhook an API Gateway wenn Job abgeschlossen ist"""
    try:
        webhook_data = {
            'jobId': job_id,
            'jobType': job_type,
            'projectId': project_id,
            'result': result,
            'status': status
        }
        
        print(f'Sende Webhook für Job {job_id} ({job_type}) an API Gateway...')
        
        response = requests.post(
            f'{API_GATEWAY_URL}/api/webhooks/job-completed',
            json=webhook_data,
            timeout=10
        )
        response.raise_for_status()
        
        print(f'Webhook für Job {job_id} erfolgreich gesendet')
        return response.json()
    except Exception as error:
        print(f'Fehler beim Senden des Webhooks für Job {job_id}: {error}')
        print(f'Webhook-URL: {API_GATEWAY_URL}/api/webhooks/job-completed')
        
        # Versuche erneut bei Verbindungsfehlern
        if hasattr(error, 'errno') and error.errno in ['ECONNREFUSED', 'ETIMEDOUT']:
            print('API Gateway nicht erreichbar, versuche Webhook erneut in 5 Sekunden...')
            time.sleep(5)
            try:
                response = requests.post(
                    f'{API_GATEWAY_URL}/api/webhooks/job-completed',
                    json=webhook_data,
                    timeout=10
                )
                response.raise_for_status()
                print(f'Webhook für Job {job_id} erfolgreich nach Wiederholung gesendet')
                return response.json()
            except Exception as retry_error:
                print(f'Webhook-Wiederholung für Job {job_id} fehlgeschlagen: {retry_error}')
        
        return None

